### PR TITLE
feat: optimize slot management

### PR DESCRIPTION
**Pull request by Tidyzq**
related #2

### CODE REVIEW

File: glass-easel/src/backend/backend_protocol.ts [hunk 1]
```diff
@@ -60,7 +61,13 @@ export interface Element {
...
   setId(id: string): void
   setSlotName(slot: string): void
-  setSlot(name: string, inherit: boolean): void
+  setContainingSlot(slot: Element | undefined | null): void
+  reassignContainingSlot(oldSlot: Element | null, newSlot: Element | null): void
+  spliceBeforeSlotNodes(before: number, deleteCount: number, list: Element): void
+  spliceAppendSlotNodes(list: Element): void
+  spliceRemoveSlotNodes(before: number, deleteCount: number): void
+  setInheritSlots(): void
+  setVirtualHost(): void
```
Comment (LastLeaf) [inline glass-easel/src/backend/backend_protocol.ts:70]: Why an independent call?

File: glass-easel/src/element.ts [hunk 2]
```diff
@@ -111,6 +119,14 @@ export class Element implements NodeCast {
...
   /** The child nodes (must not be modified directly!) */
   childNodes: Node[]
+  /** The index in parentNode.childNodes (-1 if no parentNode) (must not be modified directly!) */
+  parentIndex: number
+  /** The parent slot element in composed tree (must not be modified directly!) */
+  containingSlot: Element | null | undefined
+  /** The slot content nodes composed tree (must not be modified directly!) */
+  slotNodes: Node[] | undefined
+  /** The index in containingSlot.slotNodes (must not be modified directly!) */
+  slotIndex: number | undefined
```
Comment (LastLeaf) [inline glass-easel/src/element.ts:129]: Management overhead?

File: glass-easel/src/element.ts [hunk 3]
```diff
@@ -627,179 +837,62 @@ export class Element implements NodeCast {
...
     // detect whether it is in single-slot mode
     let sharedNonVirtualParent: composedElement | null | undefined
-    let dynamicSlotting = false
     if (slotParent instanceof Component) {
       if (slotParent._$external) {
-        const singleSlot = (slotParent.shadowRoot as ExternalShadowRoot).slot
-        sharedNonVirtualParent = singleSlot
+        sharedNonVirtualParent = (slotParent.shadowRoot as ExternalShadowRoot).slot
       } else {
         parentConverted = true
```
Comment (LastLeaf) [inline glass-easel/src/element.ts:844]: Why removes single slot optimization?

File: glass-easel/src/event.ts [hunk 4]
```diff
@@ -239,77 +239,79 @@ export class Event<TDetail> {
...
     ) => {
       const recShadow = (target: Element): Element | null => {
-        let cur = target
-        const targetCaller =
-          target instanceof Component ? (target.getMethodCaller() as Element) : target
+        let currentTarget = target
         const mark = target.collectMarks()
         for (;;) {
-          if (f(cur, targetCaller, mark) === false) return null
+          if (f(currentTarget, target, mark) === false) return null
```
Comment (LastLeaf) [inline glass-easel/src/event.ts:297]: Missing `getMethodCaller` ?

File: glass-easel/src/event.ts [hunk 5]
```diff
@@ -239,77 +239,79 @@ export class Event<TDetail> {
     return this._$eventBubblingControl.mutated
   }
 
+  callListener(
```
Comment (LastLeaf) [inline glass-easel/src/event.ts:242]: Private method? Add `private _$` if so.

File: glass-easel/src/shadow_root.ts [hunk 6]
```diff
@@ -47,23 +47,23 @@ export class ShadowRoot extends VirtualNode {
...
+  private _$slotsList?: Record<string, DoubleLinkedList<Element>>
+  private _$dynamicSlotsInserted?: boolean
+  private _$dynamicSlots?: Map<Element, AppliedSlotMeta>
   private _$requiredSlotValueNames?: string[]
   private _$propertyPassingDeepCopy?: DeepCopyStrategy
   private _$insertDynamicSlotHandler?: (
-    slot: Element,
-    name: string,
-    slotValues: { [name: string]: unknown },
+    slots: {
```
Comment (LastLeaf) [inline glass-easel/src/shadow_root.ts:60]: Why changed to an array?

File: glass-easel/src/shadow_root.ts [hunk 7]
```diff
@@ -354,27 +350,30 @@ export class ShadowRoot extends VirtualNode {
    * The provided node must be a valid child node of the host of this shadow root.
    * Otherwise the behavior is undefined.
    */
-  getContainingSlot(elem: Node | null): Element | null {
-    if (this._$singleSlot !== undefined) return this._$singleSlot
-    if (this._$dynamicSlotSet) {
+  calcContainingSlot(elem: Node | null): Element | null {
```
Comment (LastLeaf) [inline glass-easel/src/shadow_root.ts:353]: Not a good name for a public method?

File: glass-easel/src/shadow_root.ts [hunk 8]
```diff
@@ -399,307 +398,333 @@ export class ShadowRoot extends VirtualNode {
...
+    } else if (slotMode === SlotMode.Dynamic) {
+      for (let it = this._$subtreeSlotStart; it; it = it.next) {
+        if (f(it.value) === false) break
       }
     }
   }
 
   /**
-   * Iterate elements with their slots
+   * Iterate elements with their slots (slots-inherited nodes included)
```
Comment (LastLeaf) [inline glass-easel/src/shadow_root.ts:420]: Add return-value document for following several fn

File: glass-easel/src/shadow_root.ts [hunk 9]
```diff
@@ -399,307 +398,333 @@ export class ShadowRoot extends VirtualNode {
...
-          child.childNodes.forEach(recInheritedSlot)
+      return true
+    }
+    return this.forEachSlotContentInSlot((node, slot) => (slot === null ? f(node) : true))
+  }
+
+  /**
+   * Check whether a node is attached to this shadow root
+   */
+  isAttached(node: Node): boolean {
```
Comment (LastLeaf) [inline glass-easel/src/shadow_root.ts:477]: Not a good name. This is not the same as the attach status.

Consider `connectedToShadowRoot` ?

File: glass-easel/src/shadow_root.ts [hunk 10]
```diff
@@ -399,307 +398,333 @@ export class ShadowRoot extends VirtualNode {
...
-        if (child instanceof Element) {
-          if (child._$nodeSlot === name) f(child, index)
-          if (child._$inheritSlots) {
-            child.childNodes.forEach(rec)
-          }
-        } else if (name === '') {
-          f(child, index)
-        }
+      if (insertBeforeFirst) {
+        slotsList[slotName] = firstSlot.prev = { value: slot, next: firstSlot }
```
Comment (LastLeaf) [inline glass-easel/src/shadow_root.ts:504]: Missing `prev` field?

File: glass-easel/src/shadow_root.ts [hunk 11]
```diff
@@ -399,307 +398,333 @@ export class ShadowRoot extends VirtualNode {
...
+      if (!this._$dynamicSlotsInserted) return
+      const insertDynamicSlot = this._$insertDynamicSlotHandler
+      const removeDynamicSlot = this._$removeDynamicSlotHandler
+      this._$dynamicSlots!.set(slot, { updatePathTree: undefined })
+      removeDynamicSlot?.([slot])
+      insertDynamicSlot?.([{ slot, name: newName, slotValues: slot._$slotValues! }])
+    }
+  }
+
+  applySlotsInsertion(
```
Comment (LastLeaf) [inline glass-easel/src/shadow_root.ts:582]: Where is this function used?

Comment (LastLeaf) [reply glass-easel/src/shadow_root.ts:582]: The same for following methods

File: glass-easel/src/shadow_root.ts [hunk 12]
```diff
@@ -725,63 +750,216 @@ export class ShadowRoot extends VirtualNode {
...
+  static _$updateSubtreeSlotNodes(
+    parentNode: Element,
+    elements: Node[],
+    shadowRoot: ShadowRoot | null,
+    oldShadowRoot: ShadowRoot | null,
+    posIndex: number,
+  ):
+    | { updateContainingSlot: () => void; removeSlotNodes: () => void; insertSlotNodes: () => void }
+    | undefined {
+    if (!shadowRoot && !oldShadowRoot) return undefined
```
Comment (LastLeaf) [inline glass-easel/src/shadow_root.ts:826]: What about dynamic slotting?

File: glass-easel/src/tmpl/index.ts [hunk 13]
```diff
@@ -83,20 +83,20 @@ class GlassEaselTemplate implements templateEngine.Template {
     this.eventObjectFilter = c.eventObjectFilter
   }
 
-  createInstance(comp: GeneralComponent): templateEngine.TemplateInstance {
+  createInstance(comp: GeneralComponentInstance): templateEngine.TemplateInstance {
```
Comment (LastLeaf) [inline glass-easel/src/tmpl/index.ts:86]: Why change this

File: glass-easel/src/tmpl/proc_gen_wrapper.ts [hunk 14]
```diff
@@ -374,31 +372,42 @@ export class ProcGenWrapper {
...
+          childNodes.push(elem)
         }
       },
       undefined,
       undefined,
     )
+    return childNodes
+  }
+
+  handleChildrenCreationAndInsert(
```
Comment (LastLeaf) [inline glass-easel/src/tmpl/proc_gen_wrapper.ts:403]: Why combine this two steps?

File: glass-easel/tests/core/slot.test.ts
Comment (LastLeaf) [inline glass-easel/tests/core/slot.test.ts:1]: There should be some test cases to test linked list ops

File: glass-easel/src/event.ts (hunk 4 above)
Comment (Tidyzq) [reply glass-easel/src/event.ts:297]: > Missing `getMethodCaller` ?

getMethodCaller has been moved inside callListener, at
https://github.com/wechat-miniprogram/glass-easel/pull/61/files/7123570dda2a5064112a8fd681919f88c56482d8#diff-aecb916d3d1b7006122bdff2b93e6252e79281788764497e985718678e6c3c9cR258

File: glass-easel/src/event.ts (hunk 5 above)
Comment (Tidyzq) [reply glass-easel/src/event.ts:242]: > Private method? Add `private _$` if so.

It's a public method for dual-thread backend usage.

File: glass-easel/src/shadow_root.ts (hunk 9 above)
Comment (Tidyzq) [reply glass-easel/src/shadow_root.ts:477]: > Not a good name. This is not the same as the attach status.
> 
> Consider `connectedToShadowRoot` ?

Renamed to `isConnected`

File: glass-easel/src/shadow_root.ts (hunk 7 above)
Comment (Tidyzq) [reply glass-easel/src/shadow_root.ts:353]: > Not a good name for a public method?

Renamed to `getContainingSlot`

File: glass-easel/src/shadow_root.ts (hunk 10 above)
Comment (Tidyzq) [reply glass-easel/src/shadow_root.ts:504]: > Missing `prev` field?

Changed to non-optional key
```
- next?: DoubleLinkedList<T> | undefined
+ next: DoubleLinkedList<T> | null
```

File: glass-easel/src/tmpl/index.ts (hunk 13 above)
Comment (Tidyzq) [reply glass-easel/src/tmpl/index.ts]: Removed all `GeneralComponentInstance`

File: glass-easel/tests/core/slot.test.ts
Comment (Tidyzq) [reply glass-easel/tests/core/slot.test.ts:1]: Added linked slot list check in `glass-easel/tests/base/match.ts`

File: glass-easel/src/shadow_root.ts (hunk 6 above)
Comment (Tidyzq) [reply glass-easel/src/shadow_root.ts:60]: > Why changed to an array?

Changed it to array is for batch these slot operations, as batching operations can reduce backend operations and avoid complexity from growing caused by operation spreading.

File: glass-easel/src/shadow_root.ts (hunk 8 above)
Comment (Tidyzq) [reply glass-easel/src/shadow_root.ts]: > Add return-value document for following several fn

Added

File: glass-easel/src/shadow_root.ts (hunk 11 above)
Comment (Tidyzq) [reply glass-easel/src/shadow_root.ts]: > Not a public method? add `_$`

Changed to internal

File: glass-easel/src/event.ts (hunk 5 above)
Comment (LastLeaf) [reply glass-easel/src/event.ts:242]: Why?

File: glass-easel/src/shadow_root.ts (hunk 6 above)
Comment (LastLeaf) [reply glass-easel/src/shadow_root.ts:60]: But backend operations are always batched?

File: glass-easel/src/shadow_root.ts (hunk 11 above)
Comment (LastLeaf) [reply glass-easel/src/shadow_root.ts]: Have a double check on new methods?

File: glass-easel/tests/core/slot.test.ts
Comment (LastLeaf) [reply glass-easel/tests/core/slot.test.ts:1]: Could not find. Where?

File: glass-easel/src/event.ts (hunk 5 above)
Comment (Tidyzq) [reply glass-easel/src/event.ts:242]: > Why?

For a dual-thread backend, events triggered in the view thread need to bubble only in the view thread, but should call event listeners in the data thread. Therefore, data thread needs a method to directly call event listeners but not bubble.

File: glass-easel/tests/core/slot.test.ts
Comment (Tidyzq) [reply glass-easel/tests/core/slot.test.ts:1]: > Could not find. Where?

https://github.com/wechat-miniprogram/glass-easel/pull/61/files#diff-84cab81d1d196ebfdc4448ca2bb4d1ab6eec4b8269d17039a2e59e608025e3ddR111

**Pull request merged**